admiral/pkg/controller/common/rolloutcommon.go

```diff
@@ -0,0 +1,199 @@
+//Find the GTP that best matches the rollout.
+//It's assumed that the set of GTPs passed in has already been matched via the GtprolloutLabel. Now it's our job to choose the best one.
+//In order:
+// - If one and only one GTP matches the env label of the rollout - use that one. Use "default" as the default env label for all GTPs and rollout.
+// - If multiple GTPs match the rollout label, use the oldest one (Using an old one has less chance of new behavior which could impact workflows)
+//IMPORTANT: If an environment label is specified on either the GTP or the rollout, the same value must be specified on the other for them to match
+func MatchGTPsToRollout(gtpList []v1.GlobalTrafficPolicy, rollout *argo.Rollout) *v1.GlobalTrafficPolicy{
+	if rollout == nil || rollout.Name == "" {
+		log.Warn("Nil or empty GlobalTrafficPolicy provided for rollout match. Returning nil.")
+		return nil
+	}
+	rolloutEnvironment:= rollout.Spec.Template.Labels[Env]
+	if rolloutEnvironment == "" {
+		//No environment label, use default value
+		rolloutEnvironment = Default
+	}
+
+	//If one and only one GTP matches the env label of the rollout - use that one
+	if len(gtpList) == 1 {
+		gtpEnv := gtpList[0].Labels[Env]
+		if gtpEnv == "" {
+			gtpEnv = Default
+		}
+		if gtpEnv == rolloutEnvironment{
+			log.Infof("Newly added rollout with name=%v matched with GTP %v in namespace %v. Env=%v", rollout.Name, gtpList[0].Name, rollout.Namespace, gtpEnv)
+			return &gtpList[0]
+		} else {
+			return nil
+		}
+	}
+
+	if len(gtpList) == 0 {
+		return nil
+	}
+
+	var envMatchedGTPList []v1.GlobalTrafficPolicy
+
+	for _, gtp := range gtpList {
+		gtpEnv := gtp.Labels[Env]
+		if gtpEnv == "" {
+			gtpEnv = Default
+		}
+		if gtpEnv == rolloutEnvironment {
+			envMatchedGTPList = append(envMatchedGTPList, gtp)
+		}
+	}
+
+	//if one matches the environment from the gtp, return it
+	if len(envMatchedGTPList) == 1 {
+		log.Infof("Newly added rollout with name=%v matched with GTP %v in namespace %v. Env=%v", rollout.Name, envMatchedGTPList[0].Name, rollout.Namespace, rolloutEnvironment)
+		return &envMatchedGTPList[0]
+	}
+
+	//No GTPs matched the environment label
+	if len(envMatchedGTPList) == 0 {
+		return nil
+	}
+
+	//Using age as a tiebreak
+	sort.Slice(envMatchedGTPList, func(i, j int) bool {
+		iTime := envMatchedGTPList[i].CreationTimestamp.Nanosecond()
+		jTime := envMatchedGTPList[j].CreationTimestamp.Nanosecond()
+		return iTime<jTime
+	})
+
+	log.Warnf("Multiple GTPs found that match the rollout with name=%v in namespace %v. Using the oldest one, you may want to clean up your configs to prevent this in the future", rollout.Name, rollout.Namespace)
+	//return oldest gtp
+	log.Infof("Newly added rollout with name=%v matched with GTP %v in namespace %v. Env=%v", rollout.Name, envMatchedGTPList[0].Name, rollout.Namespace, rolloutEnvironment)
+	return &envMatchedGTPList[0]
+
+}
+
+
+
+//TODO :- Modify GetEnv to support rollouts. Remove below duplicate method
```

Comment:
Is this doable?

Reply:
2 ways to do this.
1. modify the method to accept obj type as input and identify whether rollout or deployment is sent in the method. We lose compile-time checks by doing this.
2. Add new param(of type rollout) to the method. This makes the method signature bulky.
Both approaches make the method bulky. I would like to keep the methods separate.